feat: autocommands for creating/deleting user functions (plus some refactoring)

blehrer opened the pull request:
Hi @kanedo! Firstly, thanks for getting a headstart on this project. It has been useful for me, and I'd like to contribute something back.

I wanted to limit the context for seeing user commands (`Jekyll{Post,Note,Draft,Promote}`) to situations where I could run them without causing them to create buffers representing files in non-existent directories. Consider the following usecase:

```
./blehrer.github.io
├── _site
│   ├── 404.html
│   ├── about
│   ├── blog
│   ├── LICENSE
│   └── README.md
├── blog
│   ├── 404.html
│   ├── _config.yml
│   ├── _drafts
│   ├── _notes
│   ├── _posts
│   ├── _site
│   ├── about.markdown
│   ├── assets
│   ├── Gemfile
│   ├── Gemfile.lock
│   └── index.markdown
├── LICENSE
├── README.md
└── sites.txt
```

In `./`, running `nvim -c JekyllDraft` would yield a buffer tied to the non-existent path `./_drafts` (... instead of `blog/_drafts`)

I have implemented an event listener that will either create or delete each Jekyll user_command, based on whether the current directory has a `Gemfile` with the word 'jekyll' in it. 

In the course of doing so, I ended up shuffling the existing code around a bit, writing luadoc, cleaning up some code, and and applying a stylua configuration to the code (same as what is used by [neovim/neovim](../../../neovim/neovim)). I have broken all of this into separate commits. The commits that are actually germane to the proposed feature are 362b1e0cdc248e6f0f2b8a3fb50d5036967bd9eb and 3dd7136a337236489754efb663c7f6bf4cf30ce9

To test the presence of expected autocommands:
> given: jekyll directory `$J`
> given: non-jekyll directory: `$H`
1. Starting in either `$J` or `$H`, run `nvim -V1 -c Telescope autocommands`; search for 'jekyll', expecting that you will find our new autocommand
2. In `$H`, run `nvim`; then, run any of the user_commands, expecting that all fail because there is no command named as such
3. In `$J` run `nvim`; then, run any of the user_commands, expecting that they all work
4. Starting in `$H` run `nvim -c cd $J`; then, run any of the user_commands, expecting them to work. Finally run `:cd $H`; attempt to run the user_commands, expecting that none are available

kanedo commented on lua/jekyll/init.lua:
```diff
@@ -117,5 +126,101 @@ M.promote_draft = function()
 end
 
+---module name
+local jekyll_plugin = 'jekyll'
```
This would need to change if the directory would be renamed, right? Should this be moved to the top then?

blehrer replied:
I think that's the *lua way* ... i too am pretty new to the language ;)

kanedo commented on lua/jekyll/init.lua:
```diff
@@ -15,92 +20,96 @@ local random_string = function (k)
+---@param path string
+---@param content string[]
+---@return integer the buffer number
+local create_buffer_with_name_and_content = function(path, content)
```
It's missing the `override` argument. It wasnt used so I guess we can remove the use in line 29

blehrer replied:
good catch! that must have gotten dropped somewhere when I was rebasing :/

kanedo commented on lua/jekyll/init.lua:
```diff
@@ -117,5 +126,101 @@ M.promote_draft = function()
+}
+
+---@type JekyllUserCommands
+M.user_commands = {
```
The linter complains here about 
```
Fields cannot be injected into the reference of `JekyllNvim` for `user_commands`. To do so, use `---@class` for `M`.
```
which I think is just a type annotation issue?

kanedo reviewed (commented):
Thanks @blehrer for taking the time to improve this plugin. Since this is my first time writing any lua its greatly appreciated. I added some small comments but other than that, LGTM

blehrer commented:
@kanedo thanks for the feedback. I have addressed what you wrote, and also added `modeline` comments to each of the Lua files, to further protect against future formatting issues.

kanedo reviewed (approved):
Thank you for this PR!
